src/PixiEditor.ChangeableDocument/Changeables/Graph/Nodes/CombineSeparate/SeparateColorNode.cs

```diff
@@ -19,39 +22,61 @@ public class SeparateColorNode : Node
     
     public FuncOutputProperty<Float1> A { get; }
     
-    
-    private FuncContext lastContext;
-    private Half4 lastColor;
-
     public SeparateColorNode()
     {
-        Color = CreateFuncInput<Half4>(nameof(Color), "COLOR", new Color());
         R = CreateFuncOutput<Float1>(nameof(R), "R", ctx => GetColor(ctx).R);
         G = CreateFuncOutput<Float1>(nameof(G), "G", ctx => GetColor(ctx).G);
         B = CreateFuncOutput<Float1>(nameof(B), "B", ctx => GetColor(ctx).B);
         A = CreateFuncOutput<Float1>(nameof(A), "A", ctx => GetColor(ctx).A);
+        Mode = CreateInput("Mode", "MODE", CombineSeparateColorMode.RGB);
+        Color = CreateFuncInput<Half4>(nameof(Color), "COLOR", new Color());
     }
 
     protected override Texture? OnExecute(RenderingContext context)
     {
         return null;
     }
     
-    private Half4 GetColor(FuncContext ctx)
+    private Half4 GetColor(FuncContext ctx) =>
+        Mode.Value switch
+        {
+            CombineSeparateColorMode.RGB => GetRgba(ctx),
+            CombineSeparateColorMode.HSV => GetHsva(ctx),
+            CombineSeparateColorMode.HSL => GetHsla(ctx)
+        };
+
+    private Half4 GetRgba(FuncContext ctx) => ctx.Builder.GetOrNewAttachedHalf4(this.GetHashCode(), Color.GetHashCode(), () => Color.Value(ctx));
+
+    private Half4 GetHsva(FuncContext ctx)
     {
-        Half4 target = null;
-        if (lastContext == ctx)
+        if (!ctx.HasContext && ctx.GetValue(Color) is Half4 constantColor)
         {
-            target = lastColor;
+            var variable = new Half4(string.Empty);
+            constantColor.ConstantValue.ToHsv(out float h, out float s, out float l);
+            variable.ConstantValue = new Color((byte)(h * 255), (byte)(s * 255), (byte)(l * 255), constantColor.ConstantValue.A);
+            
+            return variable;
         }
-        else
+
+        return ctx.Builder.GetOrNewAttachedHalf4(this.GetHashCode(), Color.GetHashCode(), RgbToHsvGetter);
+
+        Expression RgbToHsvGetter() => ctx.Builder.Functions.GetRgbToHsv(ctx.GetValue(Color));
+    }
+
+    private Half4 GetHsla(FuncContext ctx)
+    {
+        if (!ctx.HasContext && ctx.GetValue(Color) is Half4 constantColor)
         {
-            target = Color.Value(ctx);
+            var variable = new Half4(string.Empty);
+            constantColor.ConstantValue.ToHsl(out float h, out float s, out float l);
+            variable.ConstantValue = new Color((byte)(h * 255), (byte)(s * 255), (byte)(l * 255), constantColor.ConstantValue.A);
+            
+            return variable;
         }
```

Comment:
I kind of don't like this approach, Node shouldn't be concerned about dealing with shader builder. Originally my idea was that node writes C# "pseudocode" that gets then converted into shader, you shouldn't know that underneath you are dealing with constant values. This is a previous implementation of this node

```cs 
R = CreateFuncOutput<Float1>(nameof(R), "R", ctx => GetColor(ctx).R);
        G = CreateFuncOutput<Float1>(nameof(G), "G", ctx => GetColor(ctx).G);
        B = CreateFuncOutput<Float1>(nameof(B), "B", ctx => GetColor(ctx).B);
        A = CreateFuncOutput<Float1>(nameof(A), "A", ctx => GetColor(ctx).A);
...
 private Half4 GetColor(FuncContext ctx)
    {
        Half4 target = null;
        if (lastContext == ctx)
        {
            target = lastColor;
        }
        else
        {
            target = (Half4)ctx.GetValue(Color); 
        }
        
        lastColor = target;
        lastContext = ctx;
        return lastColor;
    }
```

With your changes I'd expect to do something like 

```cs
 private Half4 GetColor(FuncContext ctx) =>
        Mode.Value switch
        {
            CombineSeparateColorMode.RGB => GetRgba(ctx),
            CombineSeparateColorMode.HSV => GetHsva(ctx),
            CombineSeparateColorMode.HSL => GetHsla(ctx)
        };

    private Half4 GetRgba(FuncContext ctx)
    {
        return (Half4)ctx.GetValue(Color);
    }
    
    private Half4  GetHsva(FuncContext ctx)
    {
        var rgba = GetRgba(ctx);
        var hsva = ShaderMath.RgbaToHsva(rgba); // could be also ShaderColorMath or sth, just an example
        return ctx.NewHalf4(hsva);
    }
```

FuncContext is responsible for efficient "communication" between shader builder and node. So for example when you are getting value in node via FuncContext, it ensures that you are setting values via uniforms, not hardcoded values etc. You should use it in node instead of shader builder

On the other hand Shader builder is "dumb" it shouldn't do any optimizations, it just follows your orders hard

Reply:
My best advice is to follow the convention that `MathNode` does

Reply:
I don't remember why I didn't put it in FuncContext, I remember there being a issue
imma check when I get to it